File: scripts/delete-wing-drawers.js
#!/usr/bin/env node

const { spawn } = require('child_process');
const path = require('path');

// JSON-RPC client for MCP
class MCPClient {
  constructor(serverPath) {
    this.serverPath = serverPath;
    this.requestId = 0;
    this.server = null;
  }

  async connect() {
    return new Promise((resolve, reject) => {
      this.server = spawn('node', [this.serverPath]);

      let buffer = '';
      this.server.stdout.on('data', (data) => {
        buffer += data.toString();
        const lines = buffer.split('\n');
        buffer = lines.pop();

        for (const line of lines) {
          if (line.trim()) {
            try {
              const response = JSON.parse(line);
              if (response.id === 1 && response.result) {
                resolve();
              }
            } catch (e) {}
          }
        }
      });

      // Send initialize request
      const initRequest = {
        jsonrpc: '2.0',
        id: 1,
        method: 'initialize',
        params: {
          protocolVersion: '2024-11-05',
          capabilities: {},
          clientInfo: { name: 'delete-wing-drawers', version: '1.0.0' },
        },
      };

      this.server.stdin.write(JSON.stringify(initRequest) + '\n');

      setTimeout(() => resolve(), 2000);
    });
  }

  async callTool(name, args) {
    return new Promise((resolve, reject) => {
      this.requestId++;
      const request = {
        jsonrpc: '2.0',
        id: this.requestId,
        method: 'tools/call',
        params: { name, arguments: args },
      };

      let buffer = '';
      const onData = (data) => {
        buffer += data.toString();
        const lines = buffer.split('\n');
        buffer = lines.pop();

        for (const line of lines) {
          if (line.trim()) {
            try {
              const response = JSON.parse(line);
              if (response.id === this.requestId) {
                this.server.stdout.removeListener('data', onData);
                if (response.error) {
                  reject(new Error(response.error.message));
                } else {
                  resolve(response.result);
                }
              }
            } catch (e) {}
          }
        }
      };

      this.server.stdout.on('data', onData);
      this.server.stdin.write(JSON.stringify(request) + '\n');

      setTimeout(() => {
        this.server.stdout.removeListener('data', onData);
        reject(new Error('Timeout'));
      }, 10000);
    });
  }

  close() {
    if (this.server) {
      this.server.kill();
    }
  }
}

async function deleteWingDrawers(wingName) {
  if (!wingName) {
    console.error('Usage: node delete-wing-drawers.js <wing_name>');
    process.exit(1);
  }

  // Find mempalace server
  const serverPath = path.join(
    process.env.HOME,
    '.local/lib/mempalace/dist/index.js'
  );

  console.log(`Deleting all drawers from wing: ${wingName}`);
  console.log(`Using server: ${serverPath}`);

  const client = new MCPClient(serverPath);

  try {
    await client.connect();

    let totalDeleted = 0;
    const limit = 50;

    while (true) {
      console.log('Listing drawers...');
      const listResult = await client.callTool('mempalace_list_drawers', {
        wing: wingName,
        limit: limit,
      });

      const content = listResult.content[0].text;
      const data = JSON.parse(content);

      if (!data.drawers || data.drawers.length === 0) {
        console.log('No more drawers found.');
        break;
      }

      console.log(`Found ${data.drawers.length} drawers, deleting...`);

      for (const drawer of data.drawers) {
        console.log(`  Deleting: ${drawer.drawer_id}`);
        await client.callTool('mempalace_delete_drawer', {
          drawer_id: drawer.drawer_id,
        });
        totalDeleted++;
      }
    }

    console.log(`Finished. Total drawers deleted: ${totalDeleted}`);
  } catch (err) {
    console.error('Error:', err.message);
  } finally {
    client.close();
  }
}

const wingName = process.argv[2];
deleteWingDrawers(wingName);

File: scripts/delete-wing-drawers.py
#!/usr/bin/env python3
import sys
import json
import subprocess
import os

class MCPClient:
    def __init__(self, server_path):
        self.server_path = server_path
        self.request_id = 0
        self.proc = None
        
    def connect(self):
        self.proc = subprocess.Popen(
            ['node', self.server_path],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            bufsize=1
        )
        
        # Send initialize request
        init_req = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "initialize",
            "params": {
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {"name": "delete-wing-drawers", "version": "1.0.0"}
            }
        }
        self.proc.stdin.write(json.dumps(init_req) + '\n')
        self.proc.stdin.flush()
        
        # Wait for response
        response = self.proc.stdout.readline()
        return json.loads(response)
    
    def call_tool(self, name, args):
        self.request_id += 1
        req = {
            "jsonrpc": "2.0",
            "id": self.request_id,
            "method": "tools/call",
            "params": {"name": name, "arguments": args}
        }
        self.proc.stdin.write(json.dumps(req) + '\n')
        self.proc.stdin.flush()
        
        # Read response
        response_line = self.proc.stdout.readline()
        response = json.loads(response_line)
        
        if 'error' in response:
            raise Exception(response['error']['message'])
        return response['result']
    
    def close(self):
        if self.proc:
            self.proc.terminate()
            self.proc.wait()

def delete_wing_drawers(wing_name):
    if not wing_name:
        print("Usage: python3 delete-wing-drawers.py <wing_name>")
        sys.exit(1)
    
    # Find mempalace server
    server_path = os.path.expanduser('~/.local/lib/mempalace/dist/index.js')
    
    if not os.path.exists(server_path):
        print(f"Error: MemPalace server not found at {server_path}")
        sys.exit(1)
    
    print(f"Deleting all drawers from wing: {wing_name}")
    print(f"Using server: {server_path}")
    
    client = MCPClient(server_path)
    
    try:
        client.connect()
        
        total_deleted = 0
        limit = 50
        
        while True:
            print("Listing drawers...")
            result = client.call_tool('mempalace_list_drawers', {
                'wing': wing_name,
                'limit': limit
            })
            
            content = json.loads(result['content'][0]['text'])
            
            if not content.get('drawers') or len(content['drawers']) == 0:
                print("No more drawers found.")
                break
            
            print(f"Found {len(content['drawers'])} drawers, deleting...")
            for drawer in content['drawers']:
                print(f"  Deleting: {drawer['drawer_id']}")
                client.call_tool('mempalace_delete_drawer', {
                    'drawer_id': drawer['drawer_id']
                })
                total_deleted += 1
        
        print(f"Finished. Total drawers deleted: {total_deleted}")
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)
    finally:
        client.close()

if __name__ == '__main__':
    wing_name = sys.argv[1] if len(sys.argv) > 1 else None
    delete_wing_drawers(wing_name)
